### PR TITLE
Slade

### PR DESCRIPTION
**Pull request by rbeucher**
Can we merge this? @julesghub @jmansour ?

### CODE REVIEW

**Comment by julesghub** [comment]
We can, but it's still not robust. We should add a warning that's is still showing problems for deformed meshes.
On 25 Oct 2018, at 07:25, rbeucher <notifications@github.com<mailto:notifications@github.com>> wrote:

Can we merge this? @julesghub<https://github.com/julesghub> @jmansour<https://github.com/jmansour> ?

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/underworldcode/underworld2/pull/339

Commit Summary

  *   getting slade solver back
  *   SLCN with sourceTerm and NeumannBCs + api change
  *   Adding the Cosine Hill advection test
  *   Merge branch 'development' into slade
  *   GLORY - a bug fix for deformed meshes
  *   Merge branch 'development' into slade
  *   Merge branch 'development' into slade
  *   Extra info in error statement

File Changes

  *   A docs/test/2DCosineHillAdvection.ipynb<https://github.com/underworldcode/underworld2/pull/339/files#diff-0> (309)
  *   M docs/test/HeatEq_fluxBC.ipynb<https://github.com/underworldcode/underworld2/pull/339/files#diff-1> (368)
  *   M docs/test/SteadyState.ipynb<https://github.com/underworldcode/underworld2/pull/339/files#diff-2> (145)
  *   M libUnderworld/StgFEM/Utils/src/SemiLagrangianIntegrator.c<https://github.com/underworldcode/underworld2/pull/339/files#diff-3> (114)
  *   M libUnderworld/StgFEM/Utils/src/SemiLagrangianIntegrator.h<https://github.com/underworldcode/underworld2/pull/339/files#diff-4> (2)
  *   M underworld/swarm/_swarm.py<https://github.com/underworldcode/underworld2/pull/339/files#diff-5> (2)
  *   M underworld/systems/__init__.py<https://github.com/underworldcode/underworld2/pull/339/files#diff-6> (3)
  *   M underworld/systems/_advectiondiffusion.py<https://github.com/underworldcode/underworld2/pull/339/files#diff-7> (149)
  *   M underworld/systems/_energy_solver.py<https://github.com/underworldcode/underworld2/pull/339/files#diff-8> (6)
  *   M underworld/systems/_solver.py<https://github.com/underworldcode/underworld2/pull/339/files#diff-9> (3)
  *   M underworld/systems/_thermal.py<https://github.com/underworldcode/underworld2/pull/339/files#diff-10> (2)
  *   M underworld/utils/__init__.py<https://github.com/underworldcode/underworld2/pull/339/files#diff-11> (2)
  *   M underworld/utils/_meshvariable_projection.py<https://github.com/underworldcode/underworld2/pull/339/files#diff-12> (53)

Patch Links:

  *   https://github.com/underworldcode/underworld2/pull/339.patch
  *   https://github.com/underworldcode/underworld2/pull/339.diff

**Comment by arijitlaik** [comment]
I rember @VinceUW had some stability issues with slade, i'll check with him and let it be know, he was't using a deformed mesh as far as i remember.

File: underworld/systems/_advectiondiffusion.py [hunk 1]
```diff
@@ -154,7 +301,7 @@ def _setup(self):
             if isinstance( cond, uw.conditions.NeumannCondition ):
 
                 ### -VE flux because of the FEM discretisation method of the initial equation
-                negativeCond = uw.conditions.NeumannCondition( fn_flux=-1.0*cond.fn_flux,
```
Comment (jmansour) [inline underworld/systems/_advectiondiffusion.py]: We need to be a bit careful about the implications of doing this.  In particular, if we expose this function as a property so the user can reuse it (or replace it), do we exposed the negatived version?  If we do, it's confusing for the user, if we don't then we need to be careful about always applying the negative under the hood when required.  

Should we just neg it at the C level where it's applied?

File: underworld/utils/_meshvariable_projection.py [hunk 2]
```diff
@@ -11,6 +11,57 @@
 import underworld.systems.sle as sle
 import libUnderworld
 
+
+class SolveLinearSystem(_stgermain.StgCompoundComponent):
```
Comment (jmansour) [inline underworld/utils/_meshvariable_projection.py]: Should this guy live here ?   Is it a user level tool?

File: libUnderworld/StgFEM/Utils/src/SemiLagrangianIntegrator.c [hunk 3]
```diff
@@ -467,7 +482,10 @@ void BicubicInterpolator( FeVariable* feVariable, double* position, double* delt
       for( y_i = 0; y_i < 4; y_i++ )
          for( x_i = 0; x_i < 4; x_i++ ) {
             gNode_I = x_0 + x_i + ( y_0 + y_i ) * nNodes[0];
-            if( !Mesh_GlobalToDomain( feMesh, MT_VERTEX, gNode_I, &lNode_I ) ) abort();
+            if( !Mesh_GlobalToDomain( feMesh, MT_VERTEX, gNode_I, &lNode_I ) ){
+              printf("Error in BicubicInterpolator(), trying to build an interpolation to position (%g, %g) using node %d, a non domain node, in interpolator.\n", position[0], position[1], gNode_I);
```
Comment (jmansour) [inline libUnderworld/StgFEM/Utils/src/SemiLagrangianIntegrator.c]: This is better but it really needs to be a `Journal_Firewall()`.  `abort()` will just kill the kernel, and the error message won't (i think) appear in the notebook.

**Pull request merged**